TrackletAlgorithm/VMSMERouter.h

```diff
@@ -0,0 +1,210 @@
+// VMSMERouter
+
+// Sort AllStubs into smaller regions in phi, i.e. Virtual Modules (VMs).
+// By writing VMStubME memories, with each VMSMERouter correspond to one phi/AllStub region.
+// Needed for when project is split over both PFGAs
+
+#ifndef TrackletAlgorithm_VMSMERouter_h
+#define TrackletAlgorithm_VMSMERouter_h
+
+#include "ap_int.h"
+#include <cassert>
+
+#include "Constants.h"
+#include "AllStubMemory.h"
+#include "VMStubMEMemoryCM.h"
+#ifndef __SYNTHESIS__
+#ifdef CMSSW_GIT_HASH
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#else
+#include "DummyMessageLogger.h"
+#endif
+#endif
+
+/////////////////////////////////////////
+// Constants
+
+// Number of rz bin bits for ME
+constexpr unsigned kNbitsrzbinMELayer = kNbitsrzbin;
+constexpr unsigned kNbitsrzbinMEDisk = kNbitsrzbin + 1;
+
+//////////////////////////////////////
+// Functions used by the VMSMERouter
+
+// Returns the bits of phi corresponding to finephi, i.e. phi regions within a VM
+// vmbits is the number of bits for the VMs, i.e. coarse phi region. E.g. 32 VMs would use vmbits=5
+// finebits is the number of bits within the VM
+template<regionType inType>
+inline int iphivmFineBins(const typename AllStub<inType>::ASPHI phi, const int vmbits, const int finebits) {
+
+	auto finebin = (phi.range(phi.length() - vmbits - 1, phi.length() - vmbits - finebits));
+
+	return finebin;
+}
+
+// Get the corrected phi, i.e. phi at the average radius of the barrel
+// Corrected phi is used by ME memories in the barrel
+template<regionType inType>
+inline typename AllStub<inType>::ASPHI getPhiCorr(
+		const typename AllStub<inType>::ASPHI phi,
+		const typename AllStub<inType>::ASR r,
+		const typename AllStub<inType>::ASBEND bend, const int phiCorrTable[]) {
+
+	if (inType == DISKPS || inType == DISK2S) return phi; // Do nothing if disks
+
+	constexpr auto nrbins = 1 << kNbitsrzbin; // The number of bins for r
+
+	ap_uint<kNbitsrzbin> rbin = (r + (1 << (r.length() - 1))) >> (r.length() - kNbitsrzbin); // Which bin r belongs to. Note r = 0 is mid radius
+	auto index = bend * nrbins + rbin; // Index for where we find our correction value
+	auto corrValue = phiCorrTable[index]; // The amount we need to correct our phi
+
+	auto phicorr = phi - corrValue; // the corrected phi
+
+	// Check for overflow
+	if (phicorr < 0) phicorr = 0; // can't be less than 0
+	if (phicorr >= 1 << phi.length()) phicorr = (1 << phi.length()) - 1;  // can't be more than the max value
+
+	return phicorr;
+}
+
+// Returns a ME stub with all the values set
+template<class T, regionType inType, int layer, int disk>
+inline T createVMStubME(AllStub<inType> allStub,
+		const unsigned int index, bool negDisk, const int lutTable[],
+		const int phiCorrTable[], int& slot) {
```

Comment:
@mlarson02 I would suggest making `allStub` a const reference and making `negDisk` const as well:
```suggestion
inline T createVMStubME(const AllStub<inType>& allStub,
		const unsigned int index, const bool negDisk, const int lutTable[],
		const int phiCorrTable[], int& slot) {
```